scripts/west_commands/completion/west-completion.fish

```diff
@@ -195,15 +195,89 @@ function __zephyr_west_complete_help
     end
 end
 
+function __zephyr_west_get_os
+    set -l uname_output (uname)
+
+    switch $uname_output
+        case Linux
+            echo 'linux'
+        case CYGWIN 'MSYS*'
+            echo 'windows'
+        case Darwin
+            echo 'macos'
+        case '*'
+            echo ''
+    end
+end
+
+function __zephyr_west_get_cache_dir
+    switch (__zephyr_west_get_os)
+        case linux
+            echo "$HOME/.cache/west"
+        case windows
+            echo "$HOME/.cache/west"
+        case macos
+            echo "$HOME/Library/Caches/west"
+        case '*'
+            echo "$HOME/.cache/west"
+    end
+end
+
 function __zephyr_west_complete_board
-    set -l boards (west 2>/dev/null boards --format="{name}|{qualifiers}")
-    for board in $boards
-        set -l b (string split "|" $board)
-        set -l qualifiers (string split "," $b[2])
-        for i in $qualifiers
-            printf "%s\n" $b[1]/$i
+    set -l one_week_in_s 604800
+    set -l is_cache_valid 0 # 0: invalid; 1: valid
+
+    set -l cache_folder (__zephyr_west_get_cache_dir)
+    set -l cache_file $cache_folder/fish_boards_completion.cache
+
+    set -l zephyr_sha (git --work-tree (west topdir)/zephyr log -1 --pretty=format:'%H')
```

Comment:
`west topdir` is faster than a complete `list_boards`, roughly 1/3 of the time compared to list_boards when in tree in Zephyr. The time used by git is insignificant in this context.

However, one risk I see wrt. relying on just the Zephyr repo is that downstream users / teams will often create custom boards as well as having a custom manifest.
And this is where a lot of work will happen and be shared, for example https://github.com/nrfconnect/sdk-nrf, a project which itself is expected to be extended further downstream.

Thus relying on the `(west topdir)/zephyr` will not catch updates to boards specified by other Zephyr modules.
Perhaps a git watch on the manifest repo is a better choice.
This will catch commits to the manifest repo, thus including commits to the manifest file which is changed when the SHA of other projects are updated. This means that a manifest which is updated to point to a different Zephyr revision will also trigger a cache update. Thus board changes in both Zephyr and other projects are handled.

One issue which must be considered in such approach is
Manifest path is not required to be a git repo.
In fact a common use-case is to locally create and manifest file with just a subset of projects.
However, such local manifest folders are usually just containing the manifest file and not code.
So it can be solved an md5 or similar over the manifest file, when manifest path is not a git repo.
Thus `west manifest --path` instead of `west topdir`.
That will also address cases where a manifest is locally modified before a `west update`, when a different revision of a given project is desired.

If using `west manifest --path` in a cache solution, then an `md5sum` over path value (not content) can be used to create a dedicated cache per west workspace.
That would mean the one cache exists for the workspace indicated by `/projects/foo-workspace/repo/west.yml`, and another cache is created for  `/projects/bar-workspace/repo/west.yml`.

Thus both avoiding cluttering caches, as well as allow users to have multiple workspaces / switch manifest using `west config manifest.path foo` and `west config manifest.path bar` within the same workspace.

Reply:
That's a good approach, thanks for the feedback!

I updated the logic to do what you describe. Tell me if I misunderstood something in what you were describing 😀

Reply:
> I updated the logic to do what you describe. Tell me if I misunderstood something in what you were describing 😀

Thanks, initial look at the logic itself looks good 👍